PROTOCOL.md

```diff
@@ -0,0 +1,99 @@
+# GraphQL over WebSocket Protocol
+
+## Client-server communication
+
+Each message has a `type` field, which defined in the protocol of this package, as well as associated fields inside `payload` field, depending on the message type, and `id` field so the client can identify each response from the server.
+
+The structure of each client side message is:
+```typescript
+export interface OperationMessage {
+  payload?: any;
+  id?: string;
+  type: string;
+}
+```
+
+### Client -> Server
+
+#### GQL_CONNECTION_INIT
+Client sends this message after plain websocket connection to start the communication with the server
+
+This message also triggers the `onConnect` callback on the server.
+
+- `payload: Object` : optional parameters that the client specifies in `connectionParams`
+
+#### GQL_CONNECTION_TERMINATE
+Client sends this message to terminate the connection.
+
+#### GQL_START
+Client sends this message to execute GraphQL operation
+- `id: string` : The id of the GraphQL operation to start
+- `payload: Object`:
+    * `query: string` : GraphQL operation as string or parsed GraphQL document node
+    * `variables?: Object` : Object with GraphQL variables
+    * `operationName?: string` : GraphQL operation name
+    * `context?: any` : Execution context for the operation
+    
+#### GQL_STOP
+Client sends this message in order to stop a running GraphQL operation execution (for example: unsubscribe)
+- `id: string` : operation id
+    
+### Server -> Client
+
+#### GQL_CONNECTION_ERROR
+The server sends this message if `onConnect` callback returns `false` or throws an exception, and after sending this message, the server disconnects the client.
+
+It server also respond with this message in case of a parsing errors of the message (which does not disconnect the client, just ignore the message).
+
+- `payload: Object`: the server side error
+
+#### GQL_CONNECTION_ACK
+The server sends this message if `onConnect` callback returns any other value then `false`.
+
+#### GQL_DATA
+The server sends this message to transfter the GraphQL execution result from the server to the client.
+- `id: string` : ID of the operation that was successfully set up
+- `payload: Object` : 
+    * `data: any`: Execution result
+    * `errors?: Error[]` : Array of resolvers errors
+
+#### GQL_ERROR
+Server sends this message upon failing to **register** or **execute** a GraphQL operation (resolver errors are part of `GQL_DATA` message, and will be added as `errors` array)
```

Comment:
no need to state register, also, failing to execute is confusing because of the `execute` step of graphql.
this should be any error that happens in the server casued by the operation and before execute step started.